ray-curator/ray_curator/stages/reasoning/correctness_filter.py

```diff
@@ -0,0 +1,156 @@
+import asyncio
+import time
+
+import pandas as pd
+
+from ray_curator.backends.base import WorkerMetadata
+from ray_curator.stages.base import ProcessingStage
+from ray_curator.stages.filters.doc_filter import DocumentFilter
+from ray_curator.stages.reasoning.prompts import DEFAULT_GRADING_PROMPT_TEMPLATE
+from ray_curator.stages.services.model_client import AsyncLLMClient, LLMClient
+from ray_curator.tasks import DocumentBatch
+
+# Constants for magic values
+MAX_RETRY_ATTEMPTS = 3
+RETRY_SLEEP_SECONDS = 10
+
+
+class LLMBasedGrader(ProcessingStage[DocumentBatch, DocumentBatch]):
+    """
+    A stage for filtering reasoning traces based on correctness. It automatically detects whether to use
+    async (concurrent) or sync (sequential) processing based on the client type.
+    """
+
+    def __init__(  # noqa: PLR0913
+        self,
+        prompt: str,
+        client: AsyncLLMClient | LLMClient,
+        model_name: str,
+        input_problem_field: str,
+        input_attempt_field: str,
+        input_solution_field: str,
+        output_field: str,
+    ):
+        if prompt is not None:
+            self.prompt = prompt
+        else:
+            self.prompt = DEFAULT_GRADING_PROMPT_TEMPLATE
+        self.client = client
+        self.model_name = model_name
+        self.input_problem_field = input_problem_field
+        self.input_attempt_field = input_attempt_field
+        self.input_solution_field = input_solution_field
+        self.output_field = output_field
+        self.is_async_client = isinstance(client, AsyncLLMClient)
+        self.allowed_output_values = ["Yes", "No"]
+
+    @property
+    def name(self) -> str:
+        return "LLMBasedCorrectnessFilter"
```

Comment:
Based on changes in #764 we need to update this to
`self._name = "LLMBasedCorrectnessFilter"` instead of having it as a property

Reply:
Yup otherwise `with(..)` might get messed up